Make Python RPC handler does not hold module in global variable

xush6528 opened the pull request:
Summary:
# Problem

ProcessGroupAgent used in test_rpc has SIGSEGV on exiting.

# Solution

It was because the python module was unpexceted destructed twice.

Test Plan: `

Differential Revision: D17127093

mrshenli commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -4,43 +4,42 @@ namespace torch {
+  AutoGIL ag;
+  py::object module =
+      py::module::import("torch.distributed.internal_rpc_utils");
+  if (module.is(py::none())) {
```
When this can be none? And why import it again would work?

zhaojuanmao replied:
also address the comment

xush6528 replied:
Will remove this. import failure means wrong code.

mrshenli reviewed (commented):
Could you please elaborate more on why `py::object module_` was destructed twice? 

BTW @zhaojuanmao is working on some related changes that make RPC handler a singleton. Please get a review from her as well.

zhaojuanmao commented:
@xush6528 I have two diffs stacked to fix the tests and codes, do you want to wait for that? I hold to  export it to OSS now because current Multi processing test case class is still broken.  it is better to land our diffs after the test lib is fixed cleanly.

zhaojuanmao reviewed (approved):
approving it as tmp fix to unblock, #25656 and its stack need more time to debug

please make sure CI tests are green after rebasing

xush6528 committed:
Make Python RPC handler does not hold module in global variable (#25458)

Summary:
# Problem

ProcessGroupAgent used in test_rpc has SIGSEGV on exiting.

# Solution

It was because the python module was unpexceted destructed twice.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/25458

Test Plan: Run prototype tests on top of this diff.

Differential Revision: D17127093

Pulled By: xush6528

fbshipit-source-id: bd2c7e0b7ae0e2947c15ee4332821ad47469bcab
